client/src/actions/annotation.ts

```diff
@@ -0,0 +1,643 @@
+    const state = getState();
+    const { annoMatrix: prevAnnoMatrix, obsCrossfilter: prevObsCrossfilter } =
+      state;
+    if (!prevAnnoMatrix || !prevObsCrossfilter) return;
+
+    if (!isUserAnnotation(prevAnnoMatrix, categoryName)) {
+      throw new Error("not a user annotation");
+    }
+
+    const trimmedLabel = labelName.trim();
+    if (!trimmedLabel) return;
+
+    let obsCrossfilter = prevObsCrossfilter.addObsAnnoCategory(
+      categoryName,
+      trimmedLabel
+    );
+    if (assignSelected) {
+      obsCrossfilter = await obsCrossfilter.setObsColumnValues(
+        categoryName,
+        prevObsCrossfilter.allSelectedLabels(),
+        trimmedLabel
+      );
+    }
+
+    dispatch({
+      type: "annotation: add new label to category",
+      metadataField: categoryName,
+      newLabelText: trimmedLabel,
+      assignSelectedCells: assignSelected,
+      annoMatrix: obsCrossfilter.annoMatrix,
+      obsCrossfilter,
+    });
+    dispatch({ type: "annotation: disable add new label mode" });
+  };
+
+export const annotationDeleteLabelFromCategory =
+  (categoryName: string, labelName: string): ThunkResult =>
+  async (dispatch: AppDispatch, getState: GetState) => {
+    const { annoMatrix: prevAnnoMatrix, obsCrossfilter: prevObsCrossfilter } =
+      getState();
+    if (!prevAnnoMatrix || !prevObsCrossfilter) return;
+    if (!isUserAnnotation(prevAnnoMatrix, categoryName)) {
+      throw new Error("not a user annotation");
+    }
+
+    const obsCrossfilter = await prevObsCrossfilter.removeObsAnnoCategory(
+      categoryName,
+      labelName,
+      globals.unassignedCategoryLabel
+    );
+
+    dispatch({
+      type: "annotation: delete label",
+      metadataField: categoryName,
+      label: labelName,
+      annoMatrix: obsCrossfilter.annoMatrix,
+      obsCrossfilter,
+    });
+  };
+
+export const annotationRenameLabelInCategory =
+  (
+    categoryName: string,
+    oldLabelName: string,
+    newLabelName: string
+  ): ThunkResult =>
+  async (dispatch: AppDispatch, getState: GetState) => {
+    const { annoMatrix: prevAnnoMatrix, obsCrossfilter: prevObsCrossfilter } =
+      getState();
+    if (!prevAnnoMatrix || !prevObsCrossfilter) return;
+    if (!isUserAnnotation(prevAnnoMatrix, categoryName)) {
+      throw new Error("not a user annotation");
+    }
+
+    const trimmedName = newLabelName.trim();
+    if (!trimmedName) return;
+
+    let obsCrossfilter = await prevObsCrossfilter.resetObsColumnValues(
+      categoryName,
+      oldLabelName,
+      trimmedName
+    );
+    obsCrossfilter = await obsCrossfilter.removeObsAnnoCategory(
+      categoryName,
+      oldLabelName,
+      globals.unassignedCategoryLabel
+    );
+
+    dispatch({
+      type: "annotation: label edited",
+      editedLabel: trimmedName,
+      metadataField: categoryName,
+      label: oldLabelName,
+      annoMatrix: obsCrossfilter.annoMatrix,
+      obsCrossfilter,
+    });
+  };
+
+export const annotationLabelCurrentSelection =
```

Comment:
`annotationLabelCurrentSelection` wiring stays async to respect crossfilter semantics; helpful for reviewers mapping UI button clicks to thunks.